### PR TITLE
Gather all test-related into test mod.

### PR DESCRIPTION
**Pull request by ChengYueJia**
### Issue Link

* First mentioned in #953 

* Pr link
This is the second pr for this. The older one #1003  is abandoned as there are too much conflicts. 

### Contents
* Gather test related into test mod.
* Gather most of the `impl Circuit` into `dev.mod`
   This is for distinct the `test-circuits` and the test features. #1144

### CODE REVIEW

**Comment by ChengYueJia** [comment]
Hi, @aguzmant103 @leolara . 
This is a new pr to take a place of  https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1003#issuecomment-1451791513 , for there are lots of conflicts.
So, please have a quick review.

**Comment by leolara** [comment]
@ChengYueJia why are you updating the versions of so many dependencies?

**Comment by ChengYueJia** [comment]
> @ChengYueJia why are you updating the versions of so many dependencies?

Thanks to point it out. It's been fixed.

**Comment by leolara** [comment]
So my only question is why the rename of SuperCircuit as it seems it has the same interface than TestSuperCircuit?

**Comment by ChengYueJia** [comment]
> So my only question is why the rename of SuperCircuit as it seems it has the same interface than TestSuperCircuit?
@ Hi， Because the its `impl Circuit` in test mod, so we have to export it from test mod that the bench/tests can use the `impl Circuit` methods. Rename it is for distinct from the raw one.

File: circuit-benchmarks/src/super_circuit.rs
Comment (ChengYueJia) [reply circuit-benchmarks/src/super_circuit.rs]: I've removed this one.

**Review by leolara — APPROVED** [review]

**Comment by leolara** [comment]
@han0110 do we need another review or one is ok for this refactor?

**Comment by ChengYueJia** [comment]
@leolara @aguzmant103 hi, all. I've resolved the new conflicts~~

**Comment by leolara** [comment]
@ChengYueJia I approved already, I don't know if you need someone else approval or can be merge? @aguzmant103

**Comment by pinkiebell** [comment]
A quick test with this branch introduced compilation errors in downstream zkevm-chain.
```
error[E0433]: failed to resolve: use of undeclared crate or module `mock`
  --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/copy_circuit/test.rs:17:5
   |
17 | use mock::{test_ctx::helpers::account_0_code_account_1_no_code, TestContext, MOCK_ACCOUNTS};
   |     ^^^^ use of undeclared crate or module `mock`

error[E0432]: unresolved imports `bus_mapping::evm::gen_sha3_code`, `bus_mapping::evm::MemoryKind`
 --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/copy_circuit/test.rs:9:11
  |
9 |     evm::{gen_sha3_code, MemoryKind},
  |           ^^^^^^^^^^^^^  ^^^^^^^^^^ no `MemoryKind` in `evm`
  |           |
  |           no `gen_sha3_code` in `evm`

error[E0432]: unresolved import `mock`
  --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/copy_circuit/test.rs:17:5
   |
17 | use mock::{test_ctx::helpers::account_0_code_account_1_no_code, TestContext, MOCK_ACCOUNTS};
   |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `mock`
  --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/exp_circuit/test.rs:12:5
   |
12 | use mock::TestContext;
   |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `mock`
 --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/pi_circuit/test.rs:7:5
  |
7 | use mock::{CORRECT_MOCK_TXS, MOCK_CHAIN_ID};
  |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `ethers_signers`
 --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/super_circuit/test.rs:2:5
  |
2 | use ethers_signers::{LocalWallet, Signer};
  |     ^^^^^^^^^^^^^^ use of undeclared crate or module `ethers_signers`

error[E0432]: unresolved import `mock`
 --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/super_circuit/test.rs:5:5
  |
5 | use mock::{TestContext, MOCK_CHAIN_ID};
  |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `mock`
 --> /target/cargo/git/checkouts/zkevm-circuits-0d74ca8e844a45cf/7e18fe2/zkevm-circuits/src/tx_circuit/test.rs:9:5
  |
9 | use mock::AddrOrWallet;
  |     ^^^^ help: a similar path exists: `bus_mapping::mock`
```

**Comment by leolara** [comment]
@pinkiebell I think that was expected probably

**Comment by pinkiebell** [comment]
> @pinkiebell I think that was expected probably

Regarding the TestXXCircuit yes. But the compilation fails in the zkevm-circuits crate itself before that happens. This PR should pass #1325 first. Those errors don't show up in the ci tests yet

**Comment by hero78119** [comment]
There is build failed, after adding new ci checks
```
Compiling zkevm-circuits v0.1.0 (/home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits)
error[E0433]: failed to resolve: use of undeclared crate or module `mock`
  --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/copy_circuit/test.rs:17:5
   |
17 | use mock::{test_ctx::helpers::account_0_code_account_1_no_code, TestContext, MOCK_ACCOUNTS};
   |     ^^^^ use of undeclared crate or module `mock`

error[E0432]: unresolved imports `bus_mapping::evm::gen_sha3_code`, `bus_mapping::evm::MemoryKind`
 --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/copy_circuit/test.rs:9:11
  |
9 |     evm::{gen_sha3_code, MemoryKind},
  |           ^^^^^^^^^^^^^  ^^^^^^^^^^ no `MemoryKind` in `evm`
  |           |
  |           no `gen_sha3_code` in `evm`

error[E0432]: unresolved import `mock`
  --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/copy_circuit/test.rs:17:5
   |
17 | use mock::{test_ctx::helpers::account_0_code_account_1_no_code, TestContext, MOCK_ACCOUNTS};
   |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `mock`
  --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/exp_circuit/test.rs:12:5
   |
12 | use mock::TestContext;
   |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `mock`
 --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/pi_circuit/test.rs:7:5
  |
7 | use mock::{CORRECT_MOCK_TXS, MOCK_CHAIN_ID};
  |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `ethers_signers`
 --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/super_circuit/test.rs:2:5
  |
2 | use ethers_signers::{LocalWallet, Signer};
  |     ^^^^^^^^^^^^^^ use of undeclared crate or module `ethers_signers`

error[E0432]: unresolved import `mock`
 --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/super_circuit/test.rs:5:5
  |
5 | use mock::{TestContext, MOCK_CHAIN_ID};
  |     ^^^^ help: a similar path exists: `bus_mapping::mock`

error[E0432]: unresolved import `mock`
 --> /home/runner/work/zkevm-circuits/zkevm-circuits/zkevm-circuits/src/tx_circuit/test.rs:9:5
  |
9 | use mock::AddrOrWallet;
  |     ^^^^ help: a similar path exists: `bus_mapping::mock`
```

**Comment by hero78119** [comment]
> @pinkiebell hi. When testing locally, met the error as below. Seems like need to add `cargo add --path "${GIT_ROOT}/bus-mapping" --features test` in the workflow.
> 
> ```
> error[E0432]: unresolved imports `bus_mapping::evm::gen_sha3_code`, `bus_mapping::evm::MemoryKind`
>  --> /home/cloud123/blockchain/zkevm-circuits/zkevm-circuits/src/copy_circuit/test.rs:9:11
>   |
> 9 |     evm::{gen_sha3_code, MemoryKind},
>   |           ^^^^^^^^^^^^^  ^^^^^^^^^^ no `MemoryKind` in `evm`
>   |           |
>   |           no `gen_sha3_code` in `evm`
> 
> For more information about this error, try `rustc --explain E0432`.
> error: could not compile `zkevm-circuits` due to previous error
> ```

Hi @ChengYueJia, if you add this 
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/Cargo.toml#L15 ->
`bus-mapping = { path = "../bus-mapping", features = ["test"]}` then it will pass based on my local test. (I didn't test other features after adding this)

**Comment by ChengYueJia** [comment]
> > @pinkiebell hi. When testing locally, met the error as below. Seems like need to add `cargo add --path "${GIT_ROOT}/bus-mapping" --features test` in the workflow.
> > ```
> > error[E0432]: unresolved imports `bus_mapping::evm::gen_sha3_code`, `bus_mapping::evm::MemoryKind`
> >  --> /home/cloud123/blockchain/zkevm-circuits/zkevm-circuits/src/copy_circuit/test.rs:9:11
> >   |
> > 9 |     evm::{gen_sha3_code, MemoryKind},
> >   |           ^^^^^^^^^^^^^  ^^^^^^^^^^ no `MemoryKind` in `evm`
> >   |           |
> >   |           no `gen_sha3_code` in `evm`
> > 
> > For more information about this error, try `rustc --explain E0432`.
> > error: could not compile `zkevm-circuits` due to previous error
> > ```
> 
> Hi @ChengYueJia, if you add this https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/Cargo.toml#L15 -> `bus-mapping = { path = "../bus-mapping", features = ["test"]}` then it will pass based on my local test. (I didn't test other features after adding this)

@hero78119   Yes，It'll work locally.
 
>  Seems like need to add cargo add --path "${GIT_ROOT}/bus-mapping" --features test in the workflow.

For now, the online CI/CD is missing it. SO need to add it to workflow.

**Comment by hero78119** [comment]
> > Seems like need to add cargo add --path "${GIT_ROOT}/bus-mapping" --features test in the workflow.
> 
> For now, the online CI/CD is missing it. SO need to add it to workflow.

Hi. What I means is we dont need to modify ci workflow to add this extra path setting. Bcz zkevm-circuits should resolve bus-mapping from path by its own dependencies, this logically is what Cargo dependency management should do.

I just submit another [pull-request](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1337) which exactly the same commit as yours, with one extra commit to add the line mention above. Can see this pull request pass the ci checks.

**Comment by ChengYueJia** [comment]
> > > Seems like need to add cargo add --path "${GIT_ROOT}/bus-mapping" --features test in the workflow.
> > 
> > 
> > For now, the online CI/CD is missing it. SO need to add it to workflow.
> 
> Hi. What I means is we dont need to modify ci workflow to add this extra path setting. Bcz zkevm-circuits should resolve bus-mapping from path by its own dependencies, this logically is what Cargo dependency management should do.
> 
> I just submit another [pull-request](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1337) which exactly the same commit as yours, with one extra commit to add the line mention above. Can see this pull request pass the ci checks.

Thx for insprie me. @hero78119 
As for you solution, we have to clare that the dependencies are required for the project to run. If we add test-features for bus-mapping crate, we'll import useless crate to run.
So I've gathered the `bus-mapping/test`  into `test/test-circuit` features. 


And for now, all checkes are passed. @pinkiebell @aguzmant103 Please have the last review for this pr.

File: zkevm-circuits/src/exp_circuit.rs
Comment (pinkiebell) [inline zkevm-circuits/src/exp_circuit.rs:4]: Same

File: zkevm-circuits/src/super_circuit.rs
Comment (pinkiebell) [inline zkevm-circuits/src/super_circuit.rs:54]: Same

File: zkevm-circuits/Cargo.toml
Comment (pinkiebell) [reply zkevm-circuits/Cargo.toml:56]: If this has something todo with the `mod test` etc. Then I think it makes sense to move these `cfg(any...` statements back into the test.rs files.

**Review by pinkiebell — CHANGES_REQUESTED** [review]

File: .github/workflows/test-features.yml
Comment (ChengYueJia) [reply .github/workflows/test-features.yml:74]: Fixed.

File: zkevm-circuits/src/bytecode_circuit.rs
Comment (ChengYueJia) [reply zkevm-circuits/src/bytecode_circuit.rs]: Make sense.

Comment (pinkiebell) [reply zkevm-circuits/src/bytecode_circuit.rs]: I took a look at the feature selection and recognized that it will be not easily possible to keep a distinction with `mod test `(feature=test) and only the `test-circuits` feature that only requires the circuit implementation to be exported. Basically we have to add `cfg(any(feature = "test", test))]` and `cfg(any(feature = "test", test, feature = "test-circuits"))]` selectively in new places for each test.rs file. I think that is cumbersome and to not make this further complicated I opt into removing the `test-circuits` feature and replace it with something else later. What do you think?

cc/ @ed255 Might also want to chime in

Comment (ChengYueJia) [reply zkevm-circuits/src/bytecode_circuit.rs:9]: > I took a look at the feature selection and recognized that it will be not easily possible to keep a distinction with `mod test `(feature=test) and only the `test-circuits` feature that only requires the circuit implementation to be exported. Basically we have to add `cfg(any(feature = "test", test))]` and `cfg(any(feature = "test", test, feature = "test-circuits"))]` selectively in new places for each test.rs file. 

hi, @pinkiebell  , after viewing the #1144 . I've met the same problem.
I've been sperate the `iml Circuit` part as a dev mod to have a distinct board between 'test-circuits' feature and `test` one.

> I think that is cumbersome and to not make this further complicated I opt into removing the `test-circuits` feature and replace it with something else later. What do you think?

It's better to kept the `test-circuits` features. Because the `impl circuit` part can be used in unit tests , integrated tests and benches. With test-circuits, integrated/benches or other testers can ignore the useless unit-test.

**Review by pinkiebell — APPROVED** [review]
LGTM 🚀

**Comment by leolara** [comment]
@ChengYueJia it has dubble approve, please merge asap

**Comment by hero78119** [comment]
Added to merge queue. Thanks for your contribution @ChengYueJia

**Pull request merged**
